### PR TITLE
Reworked the book/tutorial

### PR DESCRIPTION
**Pull request by pentamassiv**
A while ago I created a draft PR ( https://github.com/gtk-rs/gir/pull/1354) to improve the book. A good number of commits were done since then. I am not that familiar with git and I think now the PR is ready, so I created this PR.
A few months have passed and I didn't find the time to add the missing information to explain the whole process of how to create the bindings for pango. The overhaul contains all the information of the current version and adds some to it. I believe this version can already be merged and is already an improvement.
To recap, the improvements include:

- Restructuring the book a little bit to make it easier to find information
- Adding description on what gir is and when it can be used
- Explanation on how to set up the project folder
- Explanation on how to find .gir files
- Mentioning of various other foot guns that a beginner might encounter

A native speaker should probably proofread it.

### CODE REVIEW

**Comment by pentamassiv** [comment]
@bilelmoussaoui do you maybe have some input for this PR as well?

**Comment by bilelmoussaoui** [comment]
> @bilelmoussaoui do you maybe have some input for this PR as well?

I will review this tomorrow morning

**Comment by pentamassiv** [comment]
Do you have some feedback Bilal?

**Comment by bilelmoussaoui** [comment]
> Do you have some feedback Bilal?

Sorry for taking too long, I have been busy with finishing up the release. I will get back to this as soon as possible

**Comment by pentamassiv** [comment]
Yes, of course. I didn't want to be impolite.
Thank you very much

**Comment by pentamassiv** [comment]
Rebased

File: book/src/tutorial/generate_docs.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# Generating documentation
+And finally the last feature! Just run the following command in the folder of the safe wrapper crate:
```
Comment (bilelmoussaoui) [inline book/src/tutorial/generate_docs.md:2]: Just to be very clear, usually we don't embed the docs into the library itself for licensing reasons, so I would make it very clear that you have to do this step only on the CI used to publish the docs and mention you can use the helper script with `python3 ./generate.py --embed-docs` or `python3 ./generate.py --strip-docs` for reverting those changes.

Additionally, I feel like this chapter would be the perfect place to mention the need for making sure the `dox` feature is properly set on your `Cargo.toml` for your safe rust wrappers and that the `dox` feature avoids linkings against the library so that building the docs don't fail on docs.rs

File: book/src/tutorial/generate_docs.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+Try building the documentation and also try it with the various features you might have
+
+```sh
+cargo doc
+```
+
+Congratulations, we are done! You have successfully created the safe wrapper for a C library!
+
+You can easily publish your generated bindings and the wrapper to crates.io to allow others to use it. Publishing crates is easy but keep in mind that they need to be maintained as well. We set up the project folder in a way that easily allows sharing the code. All that is needed is to add some information to your Cargo.toml. Gir will not override them when you re-generate bindings. Easy, right? If this is your first time publishing a crate, you can find a detailed guide [here](https://doc.rust-lang.org/cargo/reference/publishing.html).
```
Comment (bilelmoussaoui) [inline book/src/tutorial/generate_docs.md:36]: Another thing we could mention here in the future, is the usage of https://gitlab.gnome.org/World/Rust/gir-rustdoc & potentially having some CI templates for docs for both Github Actions & Gitlab

File: book/src/tutorial/introduction.md [hunk 3]
```diff
@@ -1,15 +1,6 @@
 # Tutorial
+In this tutorial, we go through the basic steps needed to generate a safe wrapper for a simple C library. We start with finding the .gir files needed and end with generating the documentation. We also look at a few common errors and how to fix them. Because this can be a bit abstract sometimes, we use the creation of the bindings for pango as an example. The example continues through all of the chapters of the tutorial. If you follow along until the end, you will have generated a safe wrapper including the documentation. In case you are stuck at any point or there are other errors and you can't figure out what's going on, don't hesitate to reach us so we can give you a hand!
```
Comment (bilelmoussaoui) [inline book/src/tutorial/introduction.md:2]:
```suggestion
In this tutorial, we go through the basic steps needed to generate a safe wrapper for a simple C library. We start with finding the .gir files needed and end with generating the documentation. We also look at a few common errors and how to fix them. Because this can be a bit abstract sometimes, we use the creation of the bindings for [Pango](https://docs.gtk.org/Pango/) as an example. The example continues through all of the chapters of the tutorial. If you follow along until the end, you will have generated a safe wrapper including the documentation. In case you are stuck at any point or there are other errors and you can't figure out what's going on, don't hesitate to reach us so we can give you a hand!
```

File: book/src/tutorial/preparation.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+```sh
+touch Gir.toml
+touch pango-sys/Gir.toml
+cd ..
+```
+
+## Installing gir
+Of course we also need to download and install [gir].
+```sh
+git submodule add https://github.com/gtk-rs/gir
```
Comment (bilelmoussaoui) [inline book/src/tutorial/preparation.md:25]: I would make the submodule don't auto update by default, see https://github.com/gtk-rs/gtk4-rs/blob/master/.gitmodules#L4, otherwise anyone using your library from git will have the useless submodules checked out

**Review by bilelmoussaoui — COMMENTED** [review]
A first round of reviews, haven't had the time/energy to go through all of it yet but will get back to it in the upcoming days, thanks a lot!

File: book/src/introduction.md [hunk 5]
```diff
@@ -1,31 +1,31 @@
...
 
-It operates on 4 different modes:
+- Adding documentation:
+After the safe wrapper is created, gir can even generate the documentation for us. Use the operating mode `doc` to do so.
 
-- `sys`: is what creates the low-level FFI bindings from the supplied `*.gir` file - these are essentially direct calls in to the related C library and are typically unsafe. The resulting crate is typically appended with -sys.
 
-- `normal`: generates another crate for a layer on top of these unsafe (sys) bindings which makes them safe for use in general Rust.
-
-- `not_bound`: allows you to see the detected types/methods that will not be generated for whatever reasons.
```
[the remark this answers is not in the document]
Comment (pentamassiv) [reply book/src/introduction.md:19]: It is mentioned in lines 15 & 16 in the same file. Do you want me to rephrase it?

File: book/src/tutorial/generate_docs.md (hunk 1 above)
Comment (pentamassiv) [reply book/src/tutorial/generate_docs.md:2]: Is there somewhere I can read up on these licensing issues so that I can explain them? I don't know a lot about licensing and don't understand why  for example it would be a problem to generate a wrapper with docs for a MIT licensed library that depends on GTK4 libraries (LGPL) and their wrapper (MIT).

I added d6032f674f419b6f8326a636ee7347561d6f8b7d to address your suggestions regarding the dox feature. There already is a description in the chapter about the Cargo.toml file of the safe wrapper, but your suggestion to also add it to the chapter about the docs makes sense in case somebody skips a chapter for example.

File: book/src/tutorial/generate_docs.md (hunk 2 above)
Comment (pentamassiv) [reply book/src/tutorial/generate_docs.md]: That's a good suggestion. I was not aware of it. I opened https://github.com/gtk-rs/gir/issues/1396 for this. The PR already is fairly large and a lot to read and this is something that could be added at a later stage too. If it is okay with you, I would prefer to not include your suggested additional information in this PR?

File: book/src/tutorial/preparation.md (hunk 4 above)
Comment (pentamassiv) [reply book/src/tutorial/preparation.md:25]: Fixed in a96242c7afab139df04823ecaf180d839fda12f2

File: book/src/tutorial/generate_docs.md (hunk 2 above)
Comment (bilelmoussaoui) [reply book/src/tutorial/generate_docs.md]: sure

File: book/src/tutorial/finding_gir_files.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+For GStreamer related dependencies, follow the above mentioned steps but clone this [repo](https://gitlab.freedesktop.org/gstreamer/gir-files-rs) instead.
+
+## Other dependencies
+If you have other dependencies, you have to find the files yourself. They can often be found in the repo containing the source of your dependencies or if you have them installed, you might find them under `/usr/share/gir-1.0/` again.
+
+## Example
+We want to generate the wrapper for pango. It is related to GTK, so in order to get its .gir files, we use the recommended way. While being in the project folder `git-tutorial`, we clone the [gir-files repo](https://github.com/gtk-rs/gir-files).
+
+```sh
+git clone --depth 1 https://github.com/gtk-rs/gir-files
```
[the remark this answers is not in the document]
Comment (pentamassiv) [reply book/src/tutorial/finding_gir_files.md]: Okay, I fixed it in 0e0c745543a16b390f5304c179e363b54bd6b41e

**Comment by pentamassiv** [comment]
I made changes to address all suggestions. For a few, I need some feedback to fix it.

**Comment by pentamassiv** [comment]
@bilelmoussaoui, did you get a chance to review the rest?

File: book/src/introduction.md [hunk 7]
```diff
@@ -1,31 +1,31 @@
 # Introduction
+[gir] is a tool to automatically generate FFI bindings and a safe wrapper for a library written in C. All you need to be able to use it, is a `.gir` file containing the [GObject introspection](https://gi.readthedocs.io/en/latest/) data for the library you want to create the bindings for, as well as the `.gir` files for all its dependencies. Have a look at the tutorial if you don't know how to [find the .gir files](tutorial/finding_gir_files.md). If your library does not provide a `.gir` file, unfortunately you cannot use [gir], but maybe you can try [rust-bindgen](https://github.com/rust-lang/rust-bindgen).
```
Comment (Hofer-Julian) [inline book/src/introduction.md:2]:
```suggestion
[gir] is a tool to automatically generate safe wrappers for a library exposing a C interface with [GObject introspection](https://gi.readthedocs.io/en/latest/) information. In order to use it, you need the `.gir` file containing the introspection data for the library you want to create the bindings for, as well as the `.gir` files for all its dependencies. Have a look at the tutorial if you don't know how to [find the .gir files](tutorial/finding_gir_files.md). If your library does not provide a `.gir` file, unfortunately you cannot use [gir], but maybe you can try [rust-bindgen](https://github.com/rust-lang/rust-bindgen).
```

File: book/src/introduction.md [hunk 8]
```diff
@@ -1,31 +1,31 @@
...
-
-- `doc`: used for documentation generation
-
-[gir]: https://github.com/gtk-rs/gir
-
-## Helpers
-
-[gir] includes a wrapper script `./generator.py` that detects `Gir.toml` configurations in the current directory (or the path(s) passed on the command-line) and generates "normal" or "sys" crates for it. Alternatively --embed-docs can be passed to prepare source-code for a documentation build by moving all documentation into it. For a complete overview of available options, pass --help.
+## Regenerating the bindings and wrapper
+In order to generate the bindings and the wrapper for the first time, the above mentioned steps should be followed. When you want to regenerate the crates because e.g. the library was updated, you can simplify the process by running the helper script `./generator.py`. The script detects `Gir.toml` configurations in the current directory and sub-directories (or the paths passed on the command-line) and generates "normal" or "sys" crates for it. Alternatively `--embed-docs` can be passed to prepare source-code for a documentation built by moving all documentation into it. For a complete overview of available options, pass `--help`.
```
Comment (Hofer-Julian) [inline book/src/introduction.md:23]:
```suggestion
In order to generate the bindings and the wrapper for the first time, the above-mentioned steps should be followed. When you want to regenerate the crates because e.g. the library was updated, you can simplify the process by running the helper script `./generator.py`. The script detects `Gir.toml` configurations in the current directory and subdirectories (or the paths passed on the command-line) and generates "normal" or "sys" crates for it. Alternatively `--embed-docs` can be passed to prepare source-code for a documentation built by moving all documentation into it. For a complete overview of available options, pass `--help`.
```

File: book/src/tutorial/finding_gir_files.md [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+Have a look at the .gir file of your library. At the beginning of the file, you probably see something similar to `<include name="GObject" version="2.0"/>`. "GObject" in this case would be a dependency and you will have to find the .gir file for your dependencies as well. In most cases it will be enough to follow the next two steps of the tutorial to get all needed files.
+
+## GTK dependencies
+If your library depends on GTK libraries, the recommended way to get the `.gir` files for them is to add the [gir-files repo](https://github.com/gtk-rs/gir-files) as a submodule as well. It's the recommended way, because some of the `.gir` files included in the libraries are invalid (missing or invalid annotations for example). These errors are already fixed in the gir files from the repo. Otherwise you could use the above mentioned methods to find the files and run the [script](https://github.com/gtk-rs/gir-files/blob/master/fix.sh) to fix the `.gir` files available in the gir-files repository (and only them!). You can run it like this (at the same level of the `.gir` files you want to patch):
```
Comment (Hofer-Julian) [inline book/src/tutorial/finding_gir_files.md:7]:
```suggestion
If your library depends on GTK libraries, the recommended way to get the `.gir` files for them is to add the [gir-files repo](https://github.com/gtk-rs/gir-files) as a submodule as well. It's the recommended way, because some of the `.gir` files included in the libraries are invalid (missing or invalid annotations for example). These errors are already fixed in the gir files from the repo. Otherwise you could use the above-mentioned methods to find the files and run the [script](https://github.com/gtk-rs/gir-files/blob/master/fix.sh) to fix the `.gir` files available in the gir-files repository (and only them!). You can run it like this (at the same level of the `.gir` files you want to patch):
```

File: book/src/tutorial/high_level_rust_api.md [hunk 10]
```diff
@@ -1,278 +1,317 @@
...
+    - Otherwise, the type comes from the current crate and we just put it into the `generate` list of the `Gir.toml` file.
+- Start with the first step again
+    
+The names of the objects are not the same as the crates names. You have to use the names of the corresponding gir files.
 
-```sh
-gir -o .
-cargo build
-```
+Okay, lets go through that process for a few objects of our example.
```
Comment (Hofer-Julian) [inline book/src/tutorial/high_level_rust_api.md:308]:
```suggestion
Okay, let's go through that process for a few objects of our example.
```

File: book/src/tutorial/preparation.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
+# Preparation
+In order to install gir and nicely structure the project, there are a few things to set up.
+
+## Set up the project folder
+In order to keep the project folder nicely organized, lets create a folder where we will work in and initialize the repo. We will create two library crates. pango will contain the safe wrapper crate and because it is a wrapper for the unsafe bindings, we create the pango-sys crate within the pango crate. If no Cargo.toml file is present in the sys create, a new one will be generated, so lets be safe and delete the automatically created file before we begin. The following commands will set up the project folder as described.
```
Comment (Hofer-Julian) [inline book/src/tutorial/preparation.md:5]:
```suggestion
In order to keep the project folder nicely organized, let's create a folder where we will work in and initialize the repo. We will create two library crates. pango will contain the safe wrapper crate and because it is a wrapper for the unsafe bindings, we create the pango-sys crate within the pango crate. If no Cargo.toml file is present in the sys create, a new one will be generated, so lets be safe and delete the automatically created file before we begin. The following commands will set up the project folder as described. 
```

File: book/src/tutorial/sys_library.md [hunk 12]
```diff
@@ -1,81 +1,45 @@
...
-  |       |---- Cargo.toml
-  |       |---- Gir.toml
-  |       |---- src/
-  |               |
-  |               |---- lib.rs
-  |---- src/
-          |
-          |---- lib.rs
-```
+You can find out the values for `library` and `version` by looking at the name of the .gir file of your library. In our case it is called Pango-1.0.gir. This tells us that the `library` is Pango and the `version` is 1.0. If you don't know what value to use for `min_cfg_version`, use the same as you use for `version`. If not all `.gir` files that are needed reside in `../../gir-files/`, you can add the path to the other files by changing `girs_directories`. If for example you also have `.gir` files in the root of your project folder, change it to `girs_directories = ["../../gir-files/", "../.."]`. Because we are generating the unsafe bindings, we use the `sys` work mode.
```
Comment (Hofer-Julian) [inline book/src/tutorial/sys_library.md:30]:
```suggestion
You can find out the values for `library` and `version` by looking at the name of the .gir file of your library. In our case it is called Pango-1.0.gir. This tells us that the `library` is Pango and the `version` is 1.0. If you don't know what value to use for `min_cfg_version`, use the same as you use for `version`. If not all needed `.gir` files reside in `../../gir-files/`, you can add the path to the other files by changing `girs_directories`. If for example you also have `.gir` files in the root of your project folder, change it to `girs_directories = ["../../gir-files/", "../.."]`. Because we are generating the unsafe bindings, we use the `sys` work mode.
```

**Review by Hofer-Julian — COMMENTED** [review]
I had a quick look for spelling, not for content.

Not specific to this PR, but I would recommend adding a newline after each sentence.
It doesn't change the output, but makes diffs much more readable in my opinion.

**Comment by pentamassiv** [comment]
Rebased on the latest master

**Comment by pentamassiv** [comment]
Thank you @Hofer-Julian for the suggestions. I have applied all of them. There are currently no other pending suggestions that I am aware of.

**Review by sdroege — APPROVED** [review]

**Comment by pentamassiv** [comment]
Thank you for merging it :)

**Comment by sdroege** [comment]
Sorry for taking so long to review it

**Comment by pentamassiv** [comment]
No problem. I understand that it was a lot to read and go through and probably not of the highest priority
